sw/host/opentitanlib/src/bootstrap/primitive.rs

```diff
@@ -0,0 +1,160 @@
+                frame.header.frame_num |= Frame::EOF;
+            }
+            frame.header.flash_offset = (i * Frame::DATA_LEN) as u32;
+            frame.data[..chunk.len()].copy_from_slice(chunk);
+            frame.header.hash = frame.header_hash();
+            frames.push(frame)
+        }
+        frames
+    }
+}
+
+/// Implements the primitive bootstrap protocol.
+pub struct Primitive {
+    pub inter_frame_delay: Duration,
+    pub flash_erase_delay: Duration,
+}
+
+impl Primitive {
+    const INTER_FRAME_DELAY: Duration = Duration::from_millis(20);
+    const FLASH_ERASE_DELAY: Duration = Duration::from_millis(200);
+
+    /// Creates a new `Primitive` protocol updater from `options`.
+    pub fn new(options: &BootstrapOptions) -> Self {
+        Primitive {
+            inter_frame_delay: options.inter_frame_delay.unwrap_or(Self::INTER_FRAME_DELAY),
+            flash_erase_delay: options.flash_erase_delay.unwrap_or(Self::FLASH_ERASE_DELAY),
+        }
+    }
+}
+
+impl UpdateProtocol for Primitive {
+    /// Performs the update protocol using the `transport` with the firmware `payload`.
+    fn update(&self, spi: &dyn Target, payload: &[u8]) -> Result<()> {
+        let frames = Frame::from_payload(payload);
+
+        let mut i = 0;
+        while i < frames.len() {
+            let frame = &frames[i];
+            log::info!(
+                "Writing frame {} (offset {:x?})",
+                i,
+                frame.header.flash_offset
+            );
+
+            // Write the frame and read back the hash of the previous frame.
+            let mut prev_hash = [0u8; std::mem::size_of::<Frame>()];
+            spi.run_transaction(&mut [Transfer::Both(frame.as_bytes(), &mut prev_hash)])?;
+
+            if i == 0 {
+                // If its the first frame, there is no hash to check.
+                // We need to give the target some time to erase the flash.
+                std::thread::sleep(self.flash_erase_delay);
+                i += 1;
+            } else {
+                std::thread::sleep(self.inter_frame_delay);
+                let want_hash = frames[i - 1].frame_hash();
+                if prev_hash[0..Frame::HASH_LEN] == want_hash {
+                    i += 1;
+                } else {
+                    log::error!(
+                        "Frame hash mismatch device:{:x?} != frame:{:x?}.",
+                        &prev_hash[0..Frame::HASH_LEN],
+                        want_hash
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hex;
+
+    const SIMPLE_BIN: &[u8; 2048] = include_bytes!("simple.bin");
+
+    #[test]
+    fn test_small_binary() -> Result<()> {
+        let frames = Frame::from_payload(SIMPLE_BIN);
+
+        assert_eq!(frames[0].header.frame_num, 0);
+        assert_eq!(frames[0].header.flash_offset, 0);
+        assert_eq!(
+            hex::encode(frames[0].header.hash),
```

Comment:
Is there a particular reason you don't want to use `xxd -c` (I think that's the flag) to just generate an array?

Is this so you get a nice hex assertion?

Reply:
Yes, I'm using the `hex` library specifically to get a nice hex assertion.